academy/4-my-own-chain/gas-meter.md

```diff
@@ -39,30 +39,103 @@ Add a line that consumes the designated amount of gas in each relevant handler:
 
 1. When handling a game creation:
 
-    ```go [https://github.com/cosmos/b9-checkers-academy-draft/blob/76abedcf3ad3f4e5186435e153e6ed0d18630a73/x/checkers/keeper/msg_server_create_game.go#L41]
+    ```go [https://github.com/cosmos/b9-checkers-academy-draft/blob/4e8a82e/x/checkers/keeper/msg_server_create_game.go#L45]
     ctx.GasMeter().ConsumeGas(types.CreateGameGas, "Create game")
     ```
 
 2. When handling a move:
 
-    ```go [https://github.com/cosmos/b9-checkers-academy-draft/blob/76abedcf3ad3f4e5186435e153e6ed0d18630a73/x/checkers/keeper/msg_server_play_move.go#L90]
+    ```go [https://github.com/cosmos/b9-checkers-academy-draft/blob/4e8a82e/x/checkers/keeper/msg_server_play_move.go#L90]
     ctx.GasMeter().ConsumeGas(types.PlayMoveGas, "Play a move")
     ```
 
 3. When handling a game rejection:
 
-    ```go [https://github.com/cosmos/b9-checkers-academy-draft/blob/76abedcf3ad3f4e5186435e153e6ed0d18630a73/x/checkers/keeper/msg_server_reject_game.go#L52]
+    ```go [https://github.com/cosmos/b9-checkers-academy-draft/blob/4e8a82e/x/checkers/keeper/msg_server_reject_game.go#L52]
     ctx.GasMeter().ConsumeGas(types.RejectGameGas, "Reject game")
     ```
 
-You don't meter gas in your `EndBlock` handler because it is called by a player sending a transaction. It is instead a service rendered by the network. If you want to account for the gas cost of an expiration, you have to devise a way to pre-collect it as part of the other messages.
+You do not meter gas in your `EndBlock` handler because it is **not** called by a player sending a transaction. Instead, it is a service rendered by the network. If you want to account for the gas cost of a game expiration, you have to devise a way to pre-collect it from players as part of the other messages.
 
 <HighlightBox type="tip">
 
-Avoid calling `ConsumeGas` from within a loop. If you know the number of times your code loops, consume all the gas ahead of the loop. With this trick the transaction may run out of gas as if you had run the steps if the signer had not sent enough gas. This strategy saves computations on your node.
+Avoid calling `ConsumeGas` from within a loop. If you know the number of times your code loops, consume all the gas ahead of the loop. With this trick, the transaction may run out of gas as if you had run the steps but the signer had not sent enough gas. This strategy saves computations on your node.
```

Comment:
@xavierlepretre , I have uncertainty regarding this paragraph (but it may well reflect my inexperience rather than a concept issue). To me the phrasing implies a desired action with an undesired result. It says:

1. Consume all gas before entering a loop. (recommendation)
2. "With this trick", the transaction may run out of gas as if... (but this sounds like a bad result)

Also, could it be a good idea to express why we should "Avoid calling `ConsumeGas` from within a loop"?

Reply:
```suggestion
As part of your code optimization, avoid calling `ConsumeGas` with a fixed gas cost, for instance `k`, from within a loop on each pass. Each pass of the loop uses computation resources, for instance `c`, on each node. If you know the number of times your code loops, for instance `n` times, you know that running the full loop is going to use `n*c` computation resources. 

Now consider the case of a user who sent a transaction without enough gas. The transaction will fail anyway, but at what point will it fail?

1. If you call `ConsumeGas(k)` within the loop, the transaction will fail during one pass of the loop, say the `m`th pass. It means that the node has already used `m*c` computation resources.
2. If you call `ConsumeGas(n*k)` once before the loop, the transaction will fail right there and then, and the node will have used `0` computation resources.

Choosing the trick in point 2 improves the effectiveness of your blockchain, and potentially protects it from spam and denial-of-service attacks.
 
Additionally, at the margin, a single call to `ConsumeGas` saves a bit of computation resources of the node.
```
@AndrewB9lab Without telling you which of your 2 choices it is, I want to make sure you get it from my rewrite.

Reply:
Fully understood! Also, thanks again for noting that I can edit your suggests at this stage, very helpful :-)